### ISSUE TITLE
training getting slower and slower

### ISSUE THREAD

**Issue by bbliao** [tensorforce/tensorforce#577]
Hi, Thanks for the great work! 
I am a newbie in this framework. What I am doing is using the PPO agent to generate a path in a network. So I revised the Game_2048 example and its environment. For each episode, I randomly choose a node in the network as the start, and let the agent select the "proper" neighbor next node to generate a path. Everything goes fine, the loss is decreasing, the entropy is decreasing, and the reward is increasing. 
However, I found that the training getting slower and slower, as the figure shows. I have checked the memory and it's not increasing during the training. So I am wondering what's wrong with it?
![traing_slower](https://user-images.githubusercontent.com/5222186/54588127-15d1dd80-4a5d-11e9-9e6a-23722008bb36.png)

**Comment by bbliao** [comment]
The state and action is defined as follows,
```
    @property
    def states(self):
        states = dict()
        states['path_embedding'] = dict(shape=(self.dim,), type='float')
        states['neighbour_embedding'] = dict(shape=(NEIGHBOUR_NUM, self.dim), type='float')
        states['action_mask'] = dict(shape=(NEIGHBOUR_NUM,), type='float')
        return states

    @property
    def actions(self):
        return dict(type='int', num_actions=NEIGHBOUR_NUM)
```
and  I use the custom network according to the [blog](https://reinforce.io/blog/introduction-to-tensorforce/):
```
def tf_apply(self, x, internals, update, return_internals=False):
        path_embedding = x['path_embedding']  # 256-dim, float
        neighbour_embedding = x['neighbour_embedding']  # 27*256-dim, float
        action_mask = x['action_mask']  # 27-dim, float
        dim = 64

        # path_embedding
        path_dense_1 = tf.layers.Dense(name="path_dense_1", activation='relu', units=dim)
        path_dense_1_output = path_dense_1(path_embedding)
        path_dense_1_output_expand = tf.expand_dims(path_dense_1_output, -2)

        # neighbour_embedding
        neighbour_dense_1 = tf.layers.Dense(name="neighbour_dense_1", activation='relu', units=dim)
        neighbour_dense_1_output = neighbour_dense_1(neighbour_embedding)

        # add
        path_neighbour = tf.add(path_dense_1_output_expand, neighbour_dense_1_output)
        path_neighbour_dense_1 = tf.layers.Dense(name="path_neighbour_dense_1", activation='relu', units=dim)
        path_neighbour_dense_1_output = path_neighbour_dense_1(path_neighbour)

        origin_output = tf.reduce_mean(path_neighbour_dense_1_output, axis=-1)  # 27 * 1
        actor_output = origin_output * action_mask
        # minus max for stability
        actor_output_sm = tf.exp(actor_output - tf.reduce_max(actor_output, axis=1, keepdims=True)) * action_mask
        actor_output_sm = actor_output_sm / (tf.reduce_sum(actor_output_sm, axis=1, keepdims=True) + _epsilon)

        # Combination
        if return_internals:
            return actor_output_sm, list()
        else:
            return actor_output_sm
```

**Comment by AlexKuhnle** [comment]
Hey,
I'm not sure why this would happen. One guess was that it is just due to longer episodes since being increasingly successful, but the difference between the timestep numbers are roughly the same, so doesn't seem to be the problem. Or is it that episodes may take longer when getting more successful (maybe as part of the environment)?
Another small note: You seem to calculate a softmax manually at the end of your network. Note that this is not necessary, as it is added implicitly in the agent (so now it happens twice).

**Comment by bbliao** [comment]
@AlexKuhnle Yeah, you are right. 
The episodes may take longer when getting more successful. When the network starts to converge, the training time for each episode is almost the same.
Thank you!

**Comment by AlexKuhnle** [comment]
Cool, closing this then.
